blog/2023-07-23-openhab-4-0-release.md

```diff
@@ -0,0 +1,251 @@
+---
+layout: BlogPost
+category: blog
+title: openHAB 4.0 Release
+author: Florian Hotze and others
+date: '2023-07-23T18:00:00+02:00'
+previewimage: /uploads/summer23.jpg
+tags:
+  - news
+excerpt: >-
+  A new major release has arrived! openHAB 4 brings some exciting new features and several improvements for our users as well as our developers!
+---
+A new major release has arrived! openHAB 4 brings some exciting new features and several improvements for our users as well as our developers!
+
+With openHAB 4, we have finally reached feature parity between UI-based and file-based configurations, with UI support for both transformation and persistence configuration. Additionally, we have added support for using any of the supported script languages for transformations and profiles, and we've leveled up Units of Measurement (UoM) support.
+But that's not all!
+As quite usual for new openHAB major releases, we also upgraded to Java 17 to stay in sync with the evolving Java ecosystem.
+Java 17 brings a wealth of benefits to our user and especially our developers, including security enhancements, performance improvements and reduced memory usage as well as new language features and APIs.
+
+As this is a new major release, we have a number of breaking changes that are required for the new features and improvements.
+So, please have a look at our [**official release notes**](https://github.com/openhab/openhab-distro/releases/tag/4.0.0) to learn about all breaking changes, new add-ons, enhancements and fixes.
+
+With that being said, we again want to share some statistics that show the progress in numbers.
+
+# Activity
+
+<!-- Share some stats about PRs, LOC etc. like in https://www.openhab.org/blog/2022-06-26-openhab-3-3-release.html#activity -->
+<!-- Thank major contributors like in https://www.openhab.org/blog/2022-12-19-openhab-3-4-release.html -->
+
+# Highlights
+
+As this is a major release, there are some major (breaking) changes, but there is no single big new thing (like the introduction of MainUI in openHAB 3) to learn about.
+Instead, we have quite a large number of new features and enhancements, and we hope everyone finds something that matters to him.
+
+Therefore, please have a look at the [**release notes**](https://github.com/openhab/openhab-distro/releases/tag/4.0.0) and check out the following highlights.
+
+## Core Runtime Enhancements
+
```

Comment:
_Jan N. Klug, openHAB Maintainer_

A lot of changes are under-the-hood.
Some of them can't be seen directly, but they improve your openHAB experience.
The core support for some UI features (like managing transformations and persistence from UI) resulted in large refactoring of the underlying core components.
This not only future-proofed our code, it also resulted in performance improvements and bug fixes.
This is also true for the watch-service (the component that tracks changes to files) and the event handling in general. 

There are two other points I would like to point out: units of measurement, scripting and thing upgrades. 

### Units of Measurement (UoM)

The finally tried to make handling of units more consistent in openHAB.
In some cases the "old" code had surprising effects (e.g. changing the display unit in a state description corrupted persisted data).

I strongly recommend to re-read the [**documentation**](/docs/concepts/units-of-measurement.html) and the [**upgrade instructions**](_________how do I link to the release notes of another release?_____________).

To make a long story short: We know introduced a `unit` metadata that fixes the unit for a given item. 
This unit is assumed if a value with no unit is send to the item and the same unit is used when data is persisted.
The unit set in a state description is only used for displaying, it does no longer affect the handling of the item's state internally.

### Scripting

#### Transformation and Profile

openHAB 3.4 introduced the `SCRIPT` transformation, but a profile was still missing.
We used the introduction of the profile to also improve handling of the transformation.
With openHAB 4 you can use any scripting language you like to create transformation scripts (even in the UI, see below) and use all of these as a profile when linking channels and items.

#### Startup Behavior

With the increasing popularity of scripting languages besides openHAB's class DSL (like JavaScript and Ruby) some issues regarding the startup behavior became more obvious.
For openHAB 4 we re-worked the startup process and it finally is much more deterministic startup behavior.
The runtime now ensures that the necessary add-ons are initialized before it tries to run rules.

#### Rule Triggers

In most scripting languages information about the reason the rule triggered is available, but in the past this was limited to item events (i.e. no information was given for time-based triggers or start-level-triggers).
With openHAB 4 you have much more control: You can extract the information about the rule trigger for all trigger types.

### Thing Upgrades

If you used managed (i.e. UI configured) things in the past, you'll have noticed: when a breaking changes in a binding occurs, you have to delete the thing and re-add it, requiring a full new configuration (including channel-links).

With openHAB 4 we introduced "thing upgrades".
Add-on developers can now provide instructions to automatically upgrade things when openHAB upgrades.
You'll notice this is already the case for a lot (but unfortunately not all) breaking changes in this release.

Reply:
Thanks for providing that section! I've just pushed it.

Wrt to the link to the update instructions for UoM: I've added a link that will link the 4.0 release notes upgrade UoM section, it should work as long as no-one changes that heading name or adds another heading named "upgrading".